### PR TITLE
Feature/update prox

### LINKED ISSUE #574: Feature request: Adding `lam` to  the regularizer class
https://github.com/ott-jax/ott/blob/ea7d5b9c162146fa8e716f24fad28b89b34712d8/src/ott/geometry/regularizers.py#L35C1-L35C33

Since all regularizer functions accessible to the user have a property $\lambda$ which controls the regularization strength, would it not make sense to add it to the abstract class? Sometimes in the code it make sense to do `cost_fn.regularizer.lam` but this gives a warning in the interpreter, as it thinks not all regularizers have a property `lam`.

### PR DESCRIPTION
**Pull request by michalk8**
closes #574

### CODE REVIEW

File: src/ott/geometry/costs.py [hunk 1]
```diff
@@ -502,12 +507,23 @@ def f_h(
 
     return f_h
 
+  @property
+  def lam(self) -> float:
+    """Strength of the regularization."""
```
Comment (marcocuturi) [inline src/ott/geometry/costs.py:512]: Because this is not necessarily super intuitive, it might be worth adding some basic explanation here that `lam` is converted to `alpha` in `__init__`, and stored there subsequently?

File: src/ott/geometry/regularizers.py [hunk 2]
```diff
@@ -305,55 +335,41 @@ def tree_flatten(self):  # noqa: D102
...
-  r"""L1 norm regularizer :math:`\lambda \|x\|_1`.
-
-  Args:
-    lam: Strength of the regularizer.
-  """
-
-  def __init__(self, lam: float = 1.0):
-    super().__init__()
-    self.lam = lam
+  r"""L1 norm regularizer :math:`\|x\|_1`."""
```
Comment (marcocuturi) [inline src/ott/geometry/regularizers.py:338]: i know this was this way before, but if we're using math-scripting, maybe `:math:\ell_1` instead of L1?

File: src/ott/geometry/regularizers.py [hunk 3]
```diff
@@ -305,55 +335,41 @@ def tree_flatten(self):  # noqa: D102
...
-
-  def tree_flatten(self):  # noqa: D102
-    return (self.lam,), {}
+    return jnp.sign(v) * jax.nn.relu(jnp.abs(v) - tau)
 
 
 @jtu.register_pytree_node_class
 class L2(ProximalOperator):
-  r"""L2 norm regularizer :math:`\frac{\lambda}{2} \left<x, A^TAx\right>`.
+  r"""L2 norm regularizer :math:`\frac{1}{2} \left<x, A^TAx\right>`.
```
Comment (marcocuturi) [inline src/ott/geometry/regularizers.py:349]: same, and we might want to call this `:math:\ell_2^2` i.e. `L22`, or `SqL2`

File: src/ott/geometry/regularizers.py [hunk 4]
```diff
@@ -363,31 +379,31 @@ class STVS(ProximalOperator):
...
-    return self.lam ** 2 * jnp.sum(y + 0.5)  # make positive
+    u = jnp.arcsinh(jnp.abs(x) / (2.0 * self.gamma))
+    y = u - 0.5 * jnp.exp(-2.0 * u)
+    return self.gamma ** 2 * jnp.sum(y + 0.5)  # make positive
 
   def prox(self, v: jnp.ndarray, tau: float = 1.0) -> jnp.ndarray:  # noqa: D102
-    tmp = 1.0 - (self.lam * tau / (jnp.abs(v) + 1e-12)) ** 2
-    return jax.nn.relu(tmp) * v
+    tmp = 1.0 - ((tau * self.gamma) / (v + 1e-12)) ** 2
+    return v * jax.nn.relu(tmp)
```
Comment (marcocuturi) [inline src/ott/geometry/regularizers.py:403]: not coming from your PR, but I find this `1e-12` a bit annoying now :) 

essentially, writing `s = (tau * self.gamma)`

we have that 

```
v * relu(1 - s/ v**2) = sign(v) * |v| * relu( 1- s/v**2) = sign(v) * relu( |v| - s|v|/v**2)
= sign(v) * relu( |v| - s/|v|) 
```
the Relu is active when

```
|v|-s/|v| >0  <=> v**2 > s
```
so maybe what we could have is

```
s = (tau * self.gamma)
return jnp.where(v**2 <= s, 0, v- s/v)
```

which is kind of agreeing quite nicely with the curve.

File: tests/geometry/costs_test.py [hunk 5]
```diff
@@ -228,25 +235,23 @@ def gd_solver(
...
   @pytest.mark.parametrize("d", [5, 31, 77])
   @pytest.mark.parametrize(
-      "reg_t", [
-          regularizers.L1,
-          regularizers.L2,
-          regularizers.STVS,
+      "cost_fn",
+      [
+          costs.RegTICost(regularizers.L1(), lam=8.0),
+          costs.RegTICost(regularizers.L2(), lam=13.0),
```
Comment (marcocuturi) [inline tests/geometry/costs_test.py:243]: Maybe `SqL2` or `L22` ?

**Review by marcocuturi — APPROVED** [review]
Thanks a lot Michal!! mostly minor comments.

File: src/ott/geometry/costs.py (hunk 1 above)
Comment (michalk8) [reply src/ott/geometry/costs.py:512]: Done!

File: src/ott/geometry/regularizers.py (hunk 2 above)
Comment (michalk8) [reply src/ott/geometry/regularizers.py:338]: Done.

File: src/ott/geometry/regularizers.py (hunk 3 above)
Comment (michalk8) [reply src/ott/geometry/regularizers.py:349]: I'd keep it as `L2`, as it seems also consistent with literature (see, e.g., [here](https://icml.cc/Conferences/2004/proceedings/papers/354.pdf)).

File: src/ott/geometry/regularizers.py (hunk 4 above)
Comment (michalk8) [reply src/ott/geometry/regularizers.py:403]: Unfortunately, both produce NaNs when `v==0`, tested using
```python
jax.grad(lambda x: regularizers.STVS(1).prox(x).sum())(jnp.zeros(10))
```

**Review by pkassraie — APPROVED** [review]
This PR effectively moves the regularization parameter $\lambda$ out of the signature for the regularization function, and attributes it to the overall loss, which I think makes more sense. This is done by adding the PostComposition class and updating the signatures. All good with me.

File: src/ott/geometry/costs.py [hunk 6]
```diff
@@ -393,20 +393,25 @@ class RegTICost(TICost):
...
 
   .. math::
-    \frac{\rho}{2}\|\cdot\|_2^2 + \text{regularizer}(\cdot)
+    \frac{\rho}{2}\|\cdot\|_2^2 + \lambda \text{regularizer}\left(\cdot\right)
 
   Args:
     regularizer: Regularization function.
-    rho: Scaling factor.
+    lam: Strength of the regularization.
+    rho: Scale of the quadratic part.
```
[the remark this answers is not in the document]
Comment (marcocuturi) [reply src/ott/geometry/costs.py:401]: I agree, but this might come handy if we want to introduce a `(1-alpha) L22 + alpha Reg`, which would be more stable numerically speaking

File: src/ott/geometry/regularizers.py (hunk 3 above)
Comment (marcocuturi) [reply src/ott/geometry/regularizers.py]: indeed, this is also used here: https://pyproximal.readthedocs.io/en/stable/api/generated/pyproximal.L2.html#pyproximal.L2

but because we make the distinction between `Euclidean` and `SqEuclidean` I feel the `Sq` would be more rigorous and help maintain consistency.

File: src/ott/geometry/regularizers.py (hunk 4 above)
Comment (michalk8) [reply src/ott/geometry/regularizers.py:403]: Done!

File: tests/geometry/costs_test.py (hunk 5 above)
Comment (michalk8) [reply tests/geometry/costs_test.py:243]: Using `SqL2`!
